pinot-common/src/main/java/org/apache/pinot/common/datablock/BaseDataBlock.java

```diff
@@ -426,168 +401,202 @@ public Map<Integer, String> getExceptions() {
   }
 
-  /**
-   * Writes the metadata section to the given data output stream.
-   */
-  protected void serializeMetadata(DataOutputStream dataOutputStream)
-      throws IOException {
-    dataOutputStream.writeInt(0);
+  @Nullable
+  @Override
+  public DataBuffer getVarSizeData() {
+    return _variableSizeData;
   }
 
   /**
-   * Deserializes the metadata section from the given byte buffer.
+   * Returns the list of serialized stats.
    * <p>
-   * This is the counterpart of {@link #serializeMetadata(DataOutputStream)} and it is guaranteed that the buffer will
-   * be positioned at the start of the metadata section when this method is called.
+   * The returned list may contain nulls, which would mean that no stats were available for that stage.
    * <p>
-   * <strong>Important:</strong> It is mandatory for implementations to leave the cursor at the end of the metadata, in
-   * the exact same position as it was when {@link #serializeMetadata(DataOutputStream)} was called.
-   * <p>
-   * <strong>Important:</strong> This method will be called at the end of the BaseDataConstructor constructor to read
-   * the metadata section. This means that it will be called <strong>before</strong> the subclass have been constructor
-   * have been called. Therefore it is not possible to use any subclass fields in this method.
+   * The list itself may also be null.
    */
-  protected void deserializeMetadata(ByteBuffer buffer)
-      throws IOException {
-    buffer.getInt();
+  @Nullable
+  @Override
+  public List<DataBuffer> getStatsByStage() {
+    return Collections.emptyList();
   }
 
-  private byte[] serializeExceptions()
-      throws IOException {
-    if (_errCodeToExceptionMap.isEmpty()) {
-      return new byte[4];
-    }
-    UnsynchronizedByteArrayOutputStream byteArrayOutputStream = new UnsynchronizedByteArrayOutputStream(1024);
-    DataOutputStream dataOutputStream = new DataOutputStream(byteArrayOutputStream);
-
-    dataOutputStream.writeInt(_errCodeToExceptionMap.size());
-
-    for (Map.Entry<Integer, String> entry : _errCodeToExceptionMap.entrySet()) {
-      int key = entry.getKey();
-      String value = entry.getValue();
-      byte[] valueBytes = value.getBytes(UTF_8);
-      dataOutputStream.writeInt(key);
-      dataOutputStream.writeInt(valueBytes.length);
-      dataOutputStream.write(valueBytes);
-    }
-
-    return byteArrayOutputStream.toByteArray();
+  @Override
+  public Raw asRaw() {
+    return this;
   }
 
-  private Map<Integer, String> deserializeExceptions(ByteBuffer buffer)
-      throws IOException {
-    int numExceptions = buffer.getInt();
-    Map<Integer, String> exceptions = new HashMap<>(HashUtil.getHashMapCapacity(numExceptions));
-    for (int i = 0; i < numExceptions; i++) {
-      int errCode = buffer.getInt();
-      String errMessage = DataTableUtils.decodeString(buffer);
-      exceptions.put(errCode, errMessage);
+  @Override
+  public final boolean equals(Object o) {
```

Comment:
I don't remember where equals is used, but TBH it seems to be useful to have equality. What we can discuss is whether two data blocks with the same data, one in columnar format and the other in row format are equal or not.

Reply:
If we really accidentally compare data block content in production code, that will be very problematic

Reply:
That is what ByteBuffer does under the hood. Also, if we end up comparing data blocks in production right now, the difference shouldn't be that large:
1. If the byte buffers are pointing to different data, we will fail fast (as fast as the first different byte is compared). We would end up returning not equal as before.
2. If we are comparing the same instances, it will success fast (as we include the `==` shortcut). We would end up returning equal as before
3. If we are comparing different instances pointing to memory that contains the same bytes, it will have a linear cost and we will be returning equal instead of not equal as before.

Only in the third case we would have problems. But probably the issue there would have been that we wanted to compare references instead of content, in which case we should not be using equals. And I it is clear we shouldn't be doing that.

Reply:
I've moved this code to a new utility class called DataBlockEquals. It also includes the ability to pass any comparator and also the ability to fail with an error message in case blocks are different, which is useful in tests.

This class is pretty similar to what Apache Arrow does with [VectorEqualsVisitor](https://arrow.apache.org/docs/java/reference/org/apache/arrow/vector/compare/VectorEqualsVisitor.html)